app/api/server/v1/cloud.ts

```diff
@@ -90,3 +94,41 @@ API.v1.addRoute(
 		},
 	},
 );
+
+API.v1.addRoute(
+	'cloud.getUpgradeTabParams',
+	{ authRequired: true },
+	{
+		async get() {
+			if (!hasRole(this.userId, 'admin')) {
+				return API.v1.unauthorized();
+			}
+
+			const { workspaceRegistered } = retrieveRegistrationStatus();
+
+			const licenses = getLicenses()
+				.filter(({ valid }) => valid)
+				.map(({ license }) => license);
+
+			// find any license that has trial
+			const trialLicense = licenses.find(({ meta }) => meta?.trial);
+
+			// if at least one license isn't trial, workspace isn't considered in trial
+			const isTrial = !licenses.map(({ meta }) => meta?.trial).includes(false);
+			const hasGoldLicense = licenses.map(({ tag }) => tag?.name === 'gold').includes(true);
+			const trialEndDate = trialLicense ? format(new Date(trialLicense.expiry), 'yyyy-MM-dd') : undefined;
+
+			const hadExpiredTrials = Boolean(settings.get('Cloud_Workspace_Had_Trial'));
+
+			const upgradeTabType = getUpgradeTabType({
+				registered: workspaceRegistered,
+				hasValidLicense: licenses.length > 0,
+				hadExpiredTrials,
+				isTrial,
+				hasGoldLicense,
+			});
+
+			return API.v1.success({ tabType: upgradeTabType, trialEndDate });
```

Comment:
Move this logic to the client side using `/api/v1/licenses.get`;
You can create a new endpoint to get the `retrieveRegistrationStatus` if necessary;
That way we don't have to mix CE and EE code.